cmfpy/algs/accelerated.py

```diff
@@ -0,0 +1,73 @@
+import numpy.linalg as la
+from .base import AbstractOptimizer
+
+
+class AcceleratedOptimizer(AbstractOptimizer):
+    """
+    Accelerated Update Rules.
+
+    From Gillis, https://arxiv.org/abs/1107.5194v2.
+    """
+
+    def __init__(self, data, dims, patience=3, tol=1-5,
+                 max_inner=0, inner_thresh=0, weightW=1, weightH=1,
+                 **kwargs):
+        super().__init__(data, dims, patience=patience, tol=tol, **kwargs)
+
+        self.max_inner = max_inner
+        self.inner_thresh = inner_thresh
+        self.weightW = weightW
+        self.weightH = weightH
+
+    def update(self):
+        """
+        Updates `H` and `W` several times, stopping either when the max
+        number of iterations has been reached, or when the change in the `W`
+        or `H` is below a threshold.
+        """
+        W_hist = [self.W.copy()]
+        self.setup_W_update()
+
+        # Update once
+        self.update_W()
+        W_hist.append(self.W.copy())
+        W_init_change = la.norm(W_hist[-1] - W_hist[-2])
+
+        # Update several more times
+        for itr in range(int(self.max_inner * self.weightW)):
+            self.update_W()
+            W_hist.append(self.W.copy())
+            if (self.inner_thresh > 0 and la.norm(W_hist[-1] - W_hist[-2])
+                    < self.inner_thresh * W_init_change):
+                break
+
+        H_hist = [self.H.copy()]
+        self.setup_H_update()
+
+        # Update once
+        self.update_H()
+        H_hist.append(self.H.copy())
+        H_init_change = la.norm(H_hist[-1] - H_hist[-2])
+
+        # Update several more times
+        for itr in range(int(self.max_inner * self.weightH)):
+            self.update_H()
+            H_hist.append(self.H.copy())
+            if (self.inner_thresh > 0 and la.norm(H_hist[-1] - H_hist[-2])
+                    < self.inner_thresh * H_init_change):
+                break
```

Comment:
This code chunk is replicated twice for updating `W` and `H`. Consider refactoring it into a function.

```python

def update(self):

    ...

    # Update W
    self.setup_W_update()
    self._accelerated_update(self.W, self.update_W)

    # Update H
    self.setup_H_update()
    self._accelerated_update(self.H, self.update_H)

    ...

def _accelerated_update(self, param, update_func):

    while itr < self.max_inner:
        update_func()
        param_hist.append(param.copy())
```